HETATM record type written out

mieczyslaw opened the pull request:
Fixes #1753 (and may contribute to related #2826)

Changes made in this Pull Request:
- added HETATM in `fmt` dictionary
- conditional to write either ATOM or HETATM records based on `atom.record_type`  

Tested on a PDB structure containing both ATOM and HETATM records and produces record types as expected.

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [X] CHANGELOG updated?
 - [X] Issue raised/referenced?

pep8speaks commented:
Hello @mieczyslaw! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/coordinates/PDB.py`](https://github.com/MDAnalysis/mdanalysis/blob/e708aa760a4d872f6cc3691dbd6185da45909a89/package/MDAnalysis/coordinates/PDB.py):

> [Line 1053:80](https://github.com/MDAnalysis/mdanalysis/blob/e708aa760a4d872f6cc3691dbd6185da45909a89/package/MDAnalysis/coordinates/PDB.py#L1053): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -1044,8 +1049,11 @@ def get_attr(attrname, default):
 
-            # .. _ATOM: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#ATOM
-            self.pdbfile.write(self.fmt['ATOM'].format(**vals))
+            if atom.record_type == 'ATOM':
```
What happens here if there isn't a record type in the attributes list? We should try to capture that case and default to a sensible value (warning users that this is happening).

mieczyslaw replied:
This is my first contribution so I'm still learning the details of MDAnalysis. Can you provide me with the example when there is no atom.record_type? I thought that as being read from PDB with ATOM or HETATM in line, this goes into record_type? 

Is your suggestion to check for the attribute and if there is no attribute (or it is not ATOM/HETATM), we should write as `ATOM` and print a warning?

IAlibay replied:
> This is my first contribution so I'm still learning the details of MDAnalysis. Can you provide me with the example when there is no atom.record_type? I thought that as being read from PDB with ATOM or HETATM in line, this goes into record_type?

So as far as I'm aware, `record_type` is only set when reading a PDB file (maybe PDBQT, haven't checked). In most cases, say when reading a GRO/XTC combination, you won't get the `record_type` attribute (purely because the readers will only provide attributes for things they can see in the input files, and `record_type` is unique to PDB files).

> Is your suggestion to check for the attribute and if there is no attribute (or it is not ATOM/HETATM), we should write as ATOM and print a warning?

Yeah, probably the simplest case would be to put a try/except clause around this to catch a `NoDataError`, throw a warning and default to `ATOM` (which would mimic the current MDA behaviour in most cases). However, since this will happen __a lot__ this could actually lead to a big decrease in performance, so we'd have to benchmark to see how bad things get.

mieczyslaw replied:
Check for missing attribute added, all tests passed.

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -1029,6 +1034,7 @@ def get_attr(attrname, default):
         tempfactors = get_attr('tempfactors', 0.0)
         atomnames = get_attr('names', 'X')
 
+        no_record_type = False
```
Done

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -1044,8 +1050,21 @@ def get_attr(attrname, default):
-            # .. _ATOM: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#ATOM
-            self.pdbfile.write(self.fmt['ATOM'].format(**vals))
+            if hasattr(atom, 'record_type'):
+                if atom.record_type == 'HETATM':
```
Done

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -1044,8 +1050,21 @@ def get_attr(attrname, default):
-            self.pdbfile.write(self.fmt['ATOM'].format(**vals))
+            if hasattr(atom, 'record_type'):
+                if atom.record_type == 'HETATM':
+                    # .. _HETATM: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#HETATM
```
Done

mieczyslaw commented:
Thank you @IAlibay for your review. I addressed all your points, run the tests, and pushed new commit. Please let me know if anything else needs to be improved.

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+        u = universe5
+        outfile = str(tmpdir.join('test-mol2-to-pdb.pdb'))
+
+        expected_msg = "Found no information for attr: " \
```
same thing for test below

mieczyslaw replied:
Done

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -493,6 +495,10 @@ class PDBWriter(base.WriterBase):
     are not set (:code:`None` or :code:`np.zeros(6)`,
     see Issue #2698).
 
+    When atom's record_type attribute is present (i.e. Universe object was
```
I would probably say here:
```suggestion
    When the record_types attribute is present (e.g. Universe object was
```

The clarification for e.g. over i.e. is that a user could technically just assign record type themselves.

mieczyslaw replied:
Done

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -493,6 +495,10 @@ class PDBWriter(base.WriterBase):
     see Issue #2698).
 
+    When atom's record_type attribute is present (i.e. Universe object was
+    created by loading a PDB file), ATOM and HETATM record type keywords
```
We should  also explain that ATOM is the default output otherwise.

mieczyslaw replied:
Done

IAlibay commented on package/CHANGELOG:
```diff
@@ -19,6 +19,8 @@ The rules for this file:
   * 2.0.0
 
 Fixes
+  * Instead of using ATOM for both ATOM and HETATM, HETATM record type
```
There's an author list for version 2.0.0 above (line 17), please do add yourself there too :)

mieczyslaw replied:
Done

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -475,6 +475,8 @@ class PDBWriter(base.WriterBase):
     .. _MODEL: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#MODEL
     .. _ENDMDL: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#ENDMDL
     .. _CONECT: http://www.wwpdb.org/documentation/file-format-content/format32/sect10.html#CONECT
+    .. _ATOM: http://www.wwpdb.org/documentation/file-format-content/format32/sect9.html#ATOM
```
As things stand currently, these links don't actually do anything so we should use them to add links in the docs (see below).

mieczyslaw replied:
Done

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -493,6 +495,10 @@ class PDBWriter(base.WriterBase):
     see Issue #2698).
 
+    When atom's record_type attribute is present (i.e. Universe object was
+    created by loading a PDB file), ATOM and HETATM record type keywords
```
We can add the links here:
```suggestion
    created by loading a PDB file), ATOM_ and HETATM_ record type keywords
```

mieczyslaw replied:
Done

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -995,6 +1006,10 @@ def _write_timestep(self, ts, multiframe=False):
            An indexing issue meant it previously used the first charater (Issue #2224)
 
+        .. versionchanged:: 2.0.0
+           when only record_type attribute is present, instead of using ATOM for both ATOM
```
Done

mieczyslaw commented:
> Thanks for your work here @mieczyslaw
> 
> I realise I've added lots of comments & suggested changes, please do not let it discourage you. These are mainly "getting to know MDAnalysis" issues which will go away after a few pull requests.

I appreciate your efforts!

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -995,6 +1006,10 @@ def _write_timestep(self, ts, multiframe=False):
            An indexing issue meant it previously used the first charater (Issue #2224)
 
+        .. versionchanged:: 2.0.0
+           when only record_type attribute is present, instead of using ATOM for both ATOM
```
Done

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -1044,8 +1060,14 @@ def get_attr(attrname, default):
+            try:
+                self.pdbfile.write(self.fmt[record_types[i]].format(**vals))
+            except KeyError:
+                warnings.warn("Found '{}' for record type, but allowed types "
```
Done

mieczyslaw commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -1044,8 +1060,14 @@ def get_attr(attrname, default):
+            except KeyError:
+                warnings.warn("Found '{}' for record type, but allowed types "
+                              "are ATOM or HETATM".format(record_types[i]))
+                raise ValueError
```
Done

RMeli commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -987,13 +999,19 @@ def _write_timestep(self, ts, multiframe=False):
+           first charater (Issue #2224)
+
+        .. versionchanged:: 2.0.0
+           When only record_type attribute is present, instead of using ATOM
+           for both ATOM and HETATM, HETATM record types are properly written
+           out (Issue #1753).
```
```suggestion
           When only record_type attribute is present, instead of using ATOM_
           for both ATOM_ and HETATM_, HETATM_ record types are properly written
           out (Issue #1753).
```

mieczyslaw replied:
@RMeli, thanks! In order to add suggested changes:
1. Should I also repeat `.. _ATOM` and `.. _HETATM` in `_write_timestep` to make it available? Or keeping it at the class level is enough (so also no need to repeat `.. _MODEL` and `.. _ENDMDL`, and move `.. _NUMMDL` to the class level).
2. Should I correct older 0.7.6 entry containing `MODEL` without `_`?

IAlibay replied:
> Should I also repeat .. _ATOM and .. _HETATM in _write_timestep to make it available? Or keeping it at the class level is enough (so also no need to repeat .. _MODEL and .. _ENDMDL, and move .. _NUMMDL to the class level).

From a local build, I don't think you need to redefine _ATOM/_HETATM, however do build the docs locally and test it out, it's always a recommended thing to do whenever doing docstring changes.

> Should I correct older 0.7.6 entry containing MODEL without _?

Can do, wouldn't hurt.

RMeli replied:
Hi @mieczyslaw

1. As far as I know only one link definition is needed, so keeping it at the class level should be enough. (I think @IAlibay just confirmed this locally.)
2. I didn't want to burden you with suggested changes on documentation you did not write, but it would be great if you could fix that too while you are at it. =)

mieczyslaw replied:
Would it be OK to:
1) move the links to PDB v.3.3?
2) Move to the class level also links in `def HEADER`, `def TITLE`, `def REMARK`, etc. ?
3) `def REMARK` contains `.. _REMARK` and `.. REMARK (update)`, is the latter in use?

mieczyslaw replied:
I will keep 3.2 as it is, added missing _CRYST1 link, move to the class level, sorted out multiple REMARK, REMARKS etc.

IAlibay commented:
@mieczyslaw, I left a series of comments regarding the tests, could you address these before I re-review please?

mieczyslaw commented:
> @mieczyslaw, I left a series of comments regarding the tests, could you address these before I re-review please?

if you mean the request to add three test cases, this is already added to `testsuite/MDAnalysisTests/coordinates/test_pdb.py`. I haven't seen any comments on that above, all were related to PDB.py.

IAlibay commented:
> if you mean the request to add three test cases, this is already added to `testsuite/MDAnalysisTests/coordinates/test_pdb.py`. I haven't seen any comments on that above, all were related to PDB.py.

@mieczyslaw It's possible that they aren't showing up on your screen because after a certain number of comments github will "hide the comments" see: 
![image](https://user-images.githubusercontent.com/12460125/89525275-d769c980-d7dd-11ea-9c2b-b98c12f5e435.png)

You need to click on "load more" to get the rest of the comments. If you can't see this, then please do have a look under the "files changed" tab, there should be comments there.

mieczyslaw commented:
> You need to click on "load more" to get the rest of the comments. If you can't see this, then please do have a look under the "files changed" tab, there should be comments there.

You're right! Thanks for your patience :)

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -188,6 +189,14 @@ def universe2(self):
 
+    @pytest.fixture
+    def universe4(self):
+        return mda.Universe(PDB_HOLE, PDB_HOLE)
```
Done. I followed the user guide which mentioned two parameters, topology and trajectory (which is the same in this case).

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+        """
+
+        u = universe4
+        u_atoms = u.select_atoms("resname ETA and record_type HETATM")
```
Done

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+        u_atoms = u.select_atoms("resname ETA and record_type HETATM")
+        assert_equal(len(u_atoms), 8)
+
+        outfile = str(tmpdir.join('test-hetatm.pdb'))
```
Just followed the logic of other tests `test_conect_bonds_conect`, `test_write_occupancies` etc. 

Should we then modify all tests to use outfile fixture?

IAlibay replied:
It looks like those two tests are outside this the TestPDBWriter class, so they won't have access to the outfile fixture. In this case my recommendation is just to switch over your own tests accordingly, the rest can be left as it is (one day we'll clean up, but I think @richardjgowers has grander plans for the PDBReader/Writer which might make spending time on cleaning up not worth it right now).

mieczyslaw replied:
Done

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+
+        outfile = str(tmpdir.join('test-hetatm.pdb'))
+        u.atoms.write(outfile)
+        written = mda.Universe(outfile, outfile)
```
Done

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+        written_atoms = written.select_atoms("resname ETA and "
+                                             "record_type HETATM")
+
+        assert_equal(len(u_atoms), len(written_atoms))
```
good to know, done

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+        written_atoms = written.select_atoms("resname ETA and "
+                                             "record_type HETATM")
+
+        assert_equal(len(u_atoms), len(written_atoms))
```
Good point, done

mieczyslaw commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +455,62 @@ def test_stringio_outofrange(self, universe3):
+            u.atoms.write(outfile)
+
+        written = mda.Universe(outfile, outfile)
+        assert_equal(len(u.atoms), len(written.atoms))
```
Done

mieczyslaw commented:
@IAlibay should I regenerate documentation as a part of this PR? when I checked the guide, it is a separate repo with its own setup process. thanks!

IAlibay commented:
> @IAlibay should I regenerate documentation as a part of this PR?

All you need to do here is check that the documentation changes you made build locally and behave as expected. See: https://userguide.mdanalysis.org/1.0.0/contributing_code.html#building-the-documentation

> when I checked the guide, it is a separate repo with its own setup process.

I'm not sure what you mean here, maybe you can post the link you're referring to, so it's easier for me to let you know what is meant in the guide?

mieczyslaw commented:
> > @IAlibay should I regenerate documentation as a part of this PR?
> 
> All you need to do here is check that the documentation changes you made build locally and behave as expected. See: https://userguide.mdanalysis.org/1.0.0/contributing_code.html#building-the-documentation

done, is there any reason that all html is zipped? e.g. index.html.gz

> > when I checked the guide, it is a separate repo with its own setup process.
> 
> I'm not sure what you mean here, maybe you can post the link you're referring to, so it's easier for me to let you know what is meant in the guide?

I can't find it now anymore, it was about cloning `MDAnalysis/UserGuide`, but not the case here as we are compiling dev docs.

I have regenerated and checked the docs. 

Only CRYST1 points to v3.3 PDB standard, do you want me to move it to 3.2 (as the whole module says about compatibility with 3.2), or we can assume that it is compatible with 3.3 so I can modify all links to 3.3?

mieczyslaw commented:
@IAlibay I see that merge conflict appeared due to another author added to Authors. Do you want me to rebase my branch?

IAlibay commented:
> Only CRYST1 points to v3.3 PDB standard, do you want me to move it to 3.2 (as the whole module says about compatibility with 3.2), or we can assume that it is compatible with 3.3 so I can modify all links to 3.3?

We probably should standardise things, as far as I'm aware there's no reason not to, but it might be worth just raising an issue detailing the 3.2/3.3 discrepancy. It can probably just be fixed in another PR if it takes too long for others to look at the issue.

> I see that merge conflict appeared due to another author added to Authors.

Please do fix the merge conflict. You can just merge develop into this branch if you want (I'm going to squash merge so either way it doesn't really matter).

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +454,61 @@ def test_stringio_outofrange(self, universe3):
+        expected_msg = ("Found 'ABNORM' for record type, but allowed "
+                        "types are ATOM or HETATM")
+
+        with pytest.raises(ValueError):
```
This has changed now, you no longer raise the warning but pass the message to the ValueError. You need to match on the ValueError instead.

mieczyslaw replied:
Done, thanks for spotting this!

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_pdb.py:
```diff
@@ -446,6 +454,61 @@ def test_stringio_outofrange(self, universe3):
+                                             "record_type HETATM")
+
+        assert len(u_hetatms) == len(written_atoms), "mismatched HETATM number"
+        assert_almost_equal(u_hetatms.atoms.positions, written_atoms.atoms.positions)
```
PEP8 > 79 characters. Do have a look at the other PEP8speaks issues and see if they can be fixed.

mieczyslaw replied:
Done, apart from links to PDB

mieczyslaw commented:
> > Only CRYST1 points to v3.3 PDB standard, do you want me to move it to 3.2 (as the whole module says about compatibility with 3.2), or we can assume that it is compatible with 3.3 so I can modify all links to 3.3?
> 
> We probably should standardise things, as far as I'm aware there's no reason not to, but it might be worth just raising an issue detailing the 3.2/3.3 discrepancy. It can probably just be fixed in another PR if it takes too long for others to look at the issue.

Done, https://github.com/MDAnalysis/mdanalysis/issues/2906
 
> > I see that merge conflict appeared due to another author added to Authors.
> 
> Please do fix the merge conflict. You can just merge develop into this branch if you want (I'm going to squash merge so either way it doesn't really matter).

Done.

Thanks again for all your efforts @IAlibay !

mieczyslaw commented:
@IAlibay I suppose the best would be if you resolve AUTHORS/CHANGELOG conflicts just before merging into develop (I previously resolve it, but if we wait for too long, this will happen again). Not sure about a failing codecov/project - seems to be related to other modules I haven't modified.

IAlibay commented:
Will do @mieczyslaw, whilst we're here could you add the 3.2 -> 3.3 changes docstring changes? (I'm currently dealing with a series of Windows install issues but should have time to review & hopefully merge by the weekend).

mieczyslaw commented:
> Will do @mieczyslaw, whilst we're here could you add the 3.2 -> 3.3 changes docstring changes? (I'm currently dealing with a series of Windows install issues but should have time to review & hopefully merge by the weekend).

Done, all links checked and docs rebuilt.

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -981,19 +991,21 @@ def _write_timestep(self, ts, multiframe=False):
+           first charater (Issue #2224)
+
+        .. versionchanged:: 2.0.0
+           When only record_type attribute is present, instead of
```
This should be plural since it's the atomgroup attribute? 
Also we probably should use:
```
:attr:`record_types`
```

mieczyslaw replied:
Done.

IAlibay commented on package/MDAnalysis/coordinates/PDB.py:
```diff
@@ -493,6 +500,11 @@ class PDBWriter(base.WriterBase):
     are not set (:code:`None` or :code:`np.zeros(6)`,
     see Issue #2698).
 
+    When the record_types attribute is present (e.g. Universe object was
```
Same as other comment, we probably should use:
```
:attr:`record_types`
```

mieczyslaw replied:
Done. Also current develop merged (CHANGELOG conflict).
